h2/src/main/org/h2/command/dml/MergeUsing.java

```diff
@@ -240,44 +240,38 @@ protected void merge(Row sourceRow) {
     }
 
     private boolean isTargetRowFound() {
-        ResultInterface rows = targetMatchQuery.query(0);
-        int countTargetRowsFound = 0;
-        Value[] targetRowIdValue = null;
-
-        while (rows.next()) {
-            countTargetRowsFound++;
-            targetRowIdValue = rows.currentRow();
-
+        try (ResultInterface rows = targetMatchQuery.query(0)) {
+            if (!rows.next()) {
+                return false;
+            }
+            Value targetRowId = rows.currentRow()[0];
+            Integer number = targetRowidsRemembered.get(targetRowId);
             // throw and exception if we have processed this _ROWID_ before...
-            if (targetRowidsRemembered.containsKey(targetRowIdValue[0])) {
+            if (number != null) {
                 throw DbException.get(ErrorCode.DUPLICATE_KEY_1,
                         "Merge using ON column expression, " +
                         "duplicate _ROWID_ target record already updated, deleted or inserted:_ROWID_="
-                                + targetRowIdValue[0].toString() + ":in:"
+                                + targetRowId + ":in:"
                                 + targetTableFilter.getTable()
                                 + ":conflicting source row number:"
-                                + targetRowidsRemembered
-                                        .get(targetRowIdValue[0]));
-            } else {
-                // remember the source column values we have used before (they
-                // are the effective ON clause keys
-                // and should not be repeated
-                targetRowidsRemembered.put(targetRowIdValue[0],
-                        sourceQueryRowNumber);
+                                + number);
             }
+            // remember the source column values we have used before (they
+            // are the effective ON clause keys
+            // and should not be repeated
+            targetRowidsRemembered.put(targetRowId, sourceQueryRowNumber);
+            if (rows.next()) {
```

Comment:
calling next() twice now?

Reply:
Ah, never mind I see now

Reply:
It's a detection that no more than one row is matched by ON exression. If the second next() returns `true` an exception should be thrown.

Before this change all matched rows were fully processed and counted in a loop. After it only the first row (if any) is processed fully, other unexpected rows are only counted by `getRowCount()` or by an empty loop for lazy results.